src/debug/goDebugProvider.ts

```diff
@@ -0,0 +1,115 @@
+import * as path from "path";
+import * as vscode from "vscode";
+
+import { IDebugProvider, PortInfo, Cancellable } from "./debugProvider";
+import * as debugUtils from "./debugUtils";
+import * as extensionUtils from "../extensionUtils";
+import { Kubectl } from "../kubectl";
+import { IDockerfile } from "../docker/parser";
+
+const defaultGoDebuggerExtensionId = "golang.go";
+const defaultGoDebuggerExtension = "Debugger for Go";
+const defaultGoDebuggerConfigType = "go";
+
+const defaultGoAppPort = "8080";
+const defaultGoDebugPort = "32768";
+const fullGoDebugOptsRegExp = /^dlv debug\s+.*(--listen[\s=][^\s]+)\S*/i;
+
+export class GoDebugProvider implements IDebugProvider {
+    public getDebuggerType(): string {
+        return defaultGoDebuggerConfigType;
+    }
+
+    public async isDebuggerInstalled(): Promise<boolean> {
+        if (vscode.extensions.getExtension(defaultGoDebuggerExtensionId)) {
+            return true;
+        }
+        const answer = await vscode.window.showInformationMessage(`Go debugging requires the '${defaultGoDebuggerExtension}' extension. Would you like to install it now?`, "Install Now");
+        if (answer === "Install Now") {
+            return await extensionUtils.installVscodeExtension(defaultGoDebuggerExtensionId);
+        }
+        return false;
+    }
+
+    public async startDebugging(workspaceFolder: string, sessionName: string, port: number): Promise<boolean> {
+        const debugConfiguration = {
+            type: "go",
+            request: "attach",
+            name: sessionName,
+            hostName: "127.0.0.1",
+            port
+        };
+        const currentFolder = (vscode.workspace.workspaceFolders || []).find((folder) => folder.name === path.basename(workspaceFolder));
+        if (!currentFolder) {
+            return false;  // shouldn't happen as workspaceFolder should be one of the workspaceFolders
+        }
+        return await vscode.debug.startDebugging(currentFolder, debugConfiguration);
+    }
+
+    public isSupportedImage(baseImage: string): boolean {
+        if (!baseImage) {
+            return false;
+        }
+        return baseImage.indexOf("golang") >= 0;
+    }
+
+    public async resolvePortsFromFile(dockerfile: IDockerfile, env: {}): Promise<PortInfo | undefined> {
+        let rawDebugPortInfo: string | undefined;
+        let rawAppPortInfo: string | undefined;
+
+        // Resolve the debug port.
+        rawDebugPortInfo = await debugUtils.promptForDebugPort(defaultGoDebugPort);
```

Comment:
Good call. I kinda copied this code from the java debugger as inspiration - should I go ahead and make that change to the java debugger as well, or leave it as-is and fix that in a follow-up?

Reply:
If you are in the process of editing the code again, you can definetly go ahead and change the java part as well. If not, don't worry, i noticed that we have a lot of duplicate code in all our debug providers and it could be a perfect moment to refactor them a bit :)